lib/spack/spack/test/concretize.py

```diff
@@ -925,6 +925,24 @@ def test_compiler_constraint_with_external_package(
         assert s.external
         assert s.satisfies(expected)
 
+    @pytest.mark.regression('20976')
+    def test_compiler_in_nonbuildable_external_package(self):
+        """Check that the compiler of a non-buildable external package does not
+           spread to other dependencies"""
+        packages_yaml = {
+            'external-common-openssl': {
+                'externals': [
+                    {'spec': 'external-common-openssl@1.1.1i%gcc', 'prefix': '/usr'}
+                ],
+                'buildable': False
+            }
+        }
+        spack.config.set('packages', packages_yaml)
+
+        s = Spec("external-common-python %clang").concretized()
+        assert s.satisfies('^external-common-openssl%gcc ^external-common-gdbm%clang')
+        assert 'external-common-perl' not in [d.name for d in s.dependencies()]
```

Comment:
I think we should generalize this test a bit and pass the spec we want to concretize and the checks we want to do as fixtures (see other tests that use `@pytest.mark.parametrize` as an example or feel free to ask if something is not clear).

One test we should probably add is the concretization of just `external-common-python`. With this PR the nodes built by Spack would use GCC. I think this is not correct, but since it is not documented anywhere I'm pinging @scheibelp to see if there's agreement on this edge case.

Reply:
I don't quite understand -- what would you expect `external-common-python` to concretize to? I would have thought `external-common-python%gcc ^external-common-openssl%gcc ^external-common-gdbm%gcc` (assuming `gcc` is the default compiler).

Reply:
@mkuron Obviously I missed one part, sorry. What I tried and checked was to have `%clang` specified in the external spec. In that case I was getting:
```
external-common-python%gcc ^external-common-openssl%clang ^external-common-gdbm%gcc
```
which is expected, since we are ignoring compilers from non-buildable packages.

Reply:
Ah, yes. Before my change, in that case you might have gotten `external-common-python%gcc ^external-common-openssl%clang ^external-common-gdbm%clang`, which is quite unexpected (especially since the compiler on external-common-gdbm could change seemingly at random when more dependencies were added). So are there any tests you want me to add?

Reply:
What I would expect in that case would be to have `clang` everywhere to minimize the number of compilers used in the DAG, but was wondering if we all agree on that.

Reply:
That sounds reasonable and was what Spack did before my change. Whether that was "correct" behavior boils down to the question whether we consider the default compiler or the non-buildable external dependency's compiler more important. To restore the old behavior, I suppose I could change my patch to `lambda x: x.compiler and (is_spec_buildable(x) or spec.root.compiler is None)` -- so when no compiler is requested, we take the non-buildable external package's compiler. If that sounds reasonable to you, I'll modify my pull request accordingly and amend the unit test.

Reply:
I'll note that one issue with this test is that you have

```
  O--Pe
 /    \
Py     \
 \      \
   ------G
```

And `O` is the external, and you specify the compiler on the direct dependent (Python). I'm not sure if this is fully testing the regression you mention in https://github.com/spack/spack/pull/20976#issue-556000178 (in that case it is a transitive dependency which is external): i think a more-thorough test would be to make `gdbm` external.

Secondly, I need to investigate why this is happening in the first place: our logic generally will attempt to pull a compiler from a parent before a child, so if you explicitly specify the compiler at the root, all dependencies should use that compiler (except the externals which have one assigned).

Thirdly (and after I verify why this is happening in the first place), would it be possible for you to make use of `MockPackage` like in some of the `spec_dag.py` tests (e.g. `test_test_deptype`) rather than creating separate `builtin.mock` packages. For tests that are strictly using package classes to create dependency hierarchies, it tends to make the test more readable.

Reply:
> Secondly, I need to investigate why this is happening in the first place: our logic generally will attempt to pull a compiler from a parent before a child, 

To follow up on this, the following diff allows the first test to pass

```
diff --git a/lib/spack/spack/spec.py b/lib/spack/spack/spec.py
index 670c49171..c7e192465 100644
--- a/lib/spack/spack/spec.py
+++ b/lib/spack/spack/spec.py
@@ -2256,6 +2256,8 @@ def _expand_virtual_packages(self, concretizer):
                 # If replacement is external then trim the dependencies
                 if replacement.external:
                     if (spec._dependencies):
+                        for dep in spec.dependencies():
+                            del dep._dependents[spec.name]
                         changed = True
                         spec._dependencies = DependencyMap()
                     replacement._dependencies = DependencyMap()
```

In other words, we weren't properly disconnecting the external spec from dependencies as part of the concretization (the connections are bidirectional and we only severed in one direction). I'm inclined to say this change would be useful (since it corrects the problem more fundamentally). However, with the addition of the second test, this change is not sufficient. I think I'll have some time tomorrow to look into that more.

Reply:
The second test fails with the fix in https://github.com/spack/spack/pull/20976#discussion_r561344859 because

* The logic in `concretize_compiler` doesn't borrow information from siblings - only parents/children (this was true before this PR)
* With the change in https://github.com/spack/spack/pull/20976#discussion_r561344859, it also won't pull from openssl as a parent (since it is external), so in the case of gdbm it picks a compiler "from scratch" (without using any information from the DAG).

I think more generally it would be good to allow for the possibility of pulling from siblings (especially when parents/children do not have hints), which would get the second test to pass.